### PR TITLE
Adding support for auto generating cloudformation resources

### PR DESCRIPTION
**Pull request by elmobp**
Redone based on current changes let me know!

### CODE REVIEW

**Comment by gergnz** [comment]
@kornypoet can I have a second set of eyes please?

File: Gemfile [hunk 1]
```diff
@@ -3,6 +3,7 @@ source 'https://rubygems.org'
 gemspec
 
 gem 'json'
+gem 'ruby-beautify'
```
Comment (kornypoet) [inline Gemfile]: Up til now, we've been able to keep `cfndsl` free of outside dependencies. Is there a way to make this work without an additional gem?

File: lib/cfndsl/errors.rb [hunk 2]
```diff
@@ -7,8 +7,8 @@ def self.error(err, idx = nil)
       if idx.nil?
         @errors.push(err + "\n" + caller.join("\n") + "\n")
       else
-        m = caller[idx].match(/^.*?:\d+:/)
-        err_loc = m ? m[0] : caller[idx]
+        m = caller(idx..idx).first.match(/^.*?:\d+:/)
```
Comment (kornypoet) [inline lib/cfndsl/errors.rb]: These seem quite unrelated. Rubocop changes?

File: spec/cli_spec.rb [hunk 3]
```diff
@@ -15,6 +15,11 @@
           -b, --disable-binding            Disable binding configuration
           -s, --specification-file FILE    Location of Cloudformation Resource Specification file
           -u, --update-specification       Update the Cloudformation Resource Specification file
+          -g RESOURCE_TYPE,RESOURCE_LOGICAL_NAME,
```
Comment (kornypoet) [inline spec/cli_spec.rb]: Can you add more tests that cover the basic functionality and output of cfnlego commands?

**Comment by kornypoet** [comment]
@elmobp The code looks pretty good altogether. I've made some comments on a few pieces I have questions on. @gergnz do we want this in the main branch or in `1.0.0`?

File: Gemfile (hunk 1 above)
Comment (gergnz) [inline Gemfile]: Suggest we add it to the development group in Gemfile.

File: lib/cfndsl/errors.rb (hunk 2 above)
Comment (elmobp) [inline lib/cfndsl/errors.rb]: Yep :( fails without it

File: spec/cli_spec.rb (hunk 3 above)
Comment (elmobp) [inline spec/cli_spec.rb]: Sure

File: Gemfile (hunk 1 above)
Comment (elmobp) [inline Gemfile]: What ever is easier I tried to make it work without it but the generated templates looked a mess

**Comment by gergnz** [comment]
@kornypoet I'm happy for this to go into a pre-1.0 release. The reason I say this is while it adds functionality it doesn't potentially break the core cfndsl and/or deprecate/remove features.

File: spec/cli_spec.rb (hunk 3 above)
Comment (elmobp) [inline spec/cli_spec.rb]: Tests added

File: Gemfile (hunk 1 above)
Comment (elmobp) [inline Gemfile]: The external dep is 178 lines of code is it worth just throwing it in embedded?

**Comment by gergnz** [comment]
Can we fix the last couple of issues so checks pass?

**Comment by elmobp** [comment]
Done but what do we want to do about the external gem dep?

**Comment by kornypoet** [comment]
Thanks @elmobp
Personally, I've really clung to the "cleanliness" of a zero-outside-dependency gem. However, if we follow the precedent set in external parameters and again in stevenjack/cfndsl#328 then we are bloating the code with outside dependencies anyways, by copy-pasting. If it was possible to lift only the essential functions and keep the code import minimal (one module, 20 lines-ish) I'd say let's stick with that. But if we are importing 100's of lines across more than one file, we are really doing a disservice to the open source community by not using what others have created.

**Comment by elmobp** [comment]
Let me see if there is a nice way to compact it in eg remove things from it we dont need

**Comment by elmobp** [comment]
This last commit should do the trick

**Comment by elmobp** [comment]
@kornypoet let me know when you happy with it all and I will squash it

File: lib/cfndsl/metadata.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  # "config" => {
+  #  "files"    => {
+  #    "/cfn.ini" => {
+  #      "content" => "test"
+  #     }
+  #    }
+  #  }
+  # })
+  #
+  class MetadataDefinition < JSONable
```
Comment (kornypoet) [inline lib/cfndsl/metadata.rb]: One final comment: how did this slip into your change? 
Metadata is defined for resources here: https://github.com/stevenjack/cfndsl/blob/master/lib/cfndsl/resources.rb#L8
and for templates here: https://github.com/stevenjack/cfndsl/blob/master/lib/cfndsl/orchestration_template.rb#L11
I know some issues have cropped up recently regarding the dsl_attrs, but this PR is not the place to address them. Maybe this wasn't intended to be included?

**Comment by kornypoet** [comment]
Awesome. I've requested one final change and then I think this is good to merge.

**Comment by kornypoet** [comment]
Awesome. I've requested one final change and then I think this is good to merge.

**Comment by elmobp** [comment]
@kornypoet metadata be gone not sure how that slipped in :(

**Comment by elmobp** [comment]
Ill squash it down now gimme 10

**Comment by elmobp** [comment]
Squashed to 5 any more than that the rebase gets very upset!
